engine/dict.go

```diff
@@ -0,0 +1,394 @@
+	"fmt"
+	"io"
+	"iter"
+	"sort"
+)
+
+var (
+	errInvalidDict = errors.New("invalid dict")
+	errKeyExpected = errors.New("key expected")
+)
+
+var (
+	atomColon = NewAtom(":")
+)
+
+// Dict is a term that represents a dictionary.
+//
+// Dicts are currently represented as a compound term using the functor `dict`.
+// The first argument is the tag. The remaining arguments create an array of sorted key-value pairs.
+type Dict interface {
+	Compound
+
+	Tag() Term
+	All() iter.Seq2[Atom, Term]
+
+	Value(key Atom) (Term, bool)
+	At(i int) (Atom, Term, bool)
+	Len() int
+}
+
+type dict struct {
+	compound
+}
+
+func newDict(args []Term) (Dict, error) {
+	args, err := processArgs(args)
+	if err != nil {
+		return nil, err
+	}
+	return &dict{
+		compound: compound{
+			functor: atomDict,
+			args:    args,
+		},
+	}, nil
+}
+
+func processArgs(args []Term) ([]Term, error) {
+	if len(args) == 0 || len(args)%2 == 0 {
+		return nil, errInvalidDict
+	}
+
+	tag := args[0]
+	rest := args[1:]
+
+	kv := make(map[Atom]Term, len(rest)/2)
+	for i := 0; i < len(rest); i += 2 {
+		key, ok := rest[i].(Atom)
+		value := rest[i+1]
+		if !ok {
+			return nil, errKeyExpected
+		}
+
+		if _, exists := kv[key]; exists {
+			return nil, duplicateKeyError{key: key}
+		}
+
+		kv[key] = value
+	}
+	keys := make([]Atom, 0, len(kv))
+	for k := range kv {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i] < keys[j]
+	})
+
+	processedArgs := make([]Term, 0, len(rest))
+	processedArgs = append(processedArgs, tag)
+	for _, key := range keys {
+		processedArgs = append(processedArgs, key, kv[key])
+	}
+
+	return processedArgs, nil
+}
+
+// WriteTerm outputs the Stream to an io.Writer.
+func (d *dict) WriteTerm(w io.Writer, opts *WriteOptions, env *Env) error {
+	err := d.Tag().WriteTerm(w, opts, env)
+	if err != nil {
+		return err
+	}
+
+	_, err = w.Write([]byte("{"))
+	if err != nil {
+		return err
+	}
+
+	for i := 1; i < d.Arity(); i = i + 2 {
+		if i > 1 {
+			if _, err = w.Write([]byte(",")); err != nil {
+				return err
+			}
+		}
+		if err := d.Arg(i).WriteTerm(w, opts, env); err != nil {
+			return err
+		}
+		if _, err = w.Write([]byte(":")); err != nil {
+			return err
+		}
+		if err := d.Arg(i+1).WriteTerm(w, opts, env); err != nil {
+			return err
+		}
+	}
+
+	_, err = w.Write([]byte("}"))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Compare compares the Stream with a Term.
+func (d *dict) Compare(t Term, env *Env) int {
+	return d.compound.Compare(t, env)
+}
+
+func (d *dict) Arg(n int) Term {
+	return d.compound.Arg(n)
+}
+
+func (d *dict) Arity() int {
+	return d.compound.Arity()
+}
+
+func (d *dict) Functor() Atom {
+	return d.compound.Functor()
+}
+
+func (d *dict) Tag() Term {
+	return d.compound.Arg(0)
+}
+
+func (d *dict) Len() int {
+	return (d.Arity() - 1) / 2
+}
+
+func (d *dict) Value(key Atom) (Term, bool) {
+	for k, v := range d.All() {
+		if k == key {
+			return v, true
+		}
+	}
+
+	return nil, false
+}
```

Comment:
Sure. That what I explained in the PR description. The use of a list of terms for Dict representation was chosen because it leverages Prolog’s pattern matching and unification capabilities through its compatibility with compound terms. However, I also suggest that there is room for improvement in this approach to enhance performance.

Please open a GitHub issue to address this matter. Thx.

Reply:
Oh, it seems you also opened issue #341 on `ichiban/prolog`. Please avoid doing that. You should close it.

Reply:
I can’t do that because I’m not the owner of the repository.